R/parameter_tuning.R

```diff
@@ -0,0 +1,218 @@
+  #'
+  #' @param parameter_list A named list of parameters. Each element is either a list of values or a single value
+  #' @param n_random_samples How many samples to draw from the cross product list. If NULL, all elements will be taken
+  #'
+  #' @return A dataframe where each row contains one value from each parameter drawn from the list
+
+  # Create the cross product of all elements in the list. The column names will be taken from the
+  # names in the list
+  param_grid = expand.grid(parameter_list)
+
+  # Take a random sampling of rows if requested and the number is lower than the number of rows
+  if (!is.null(n_random_samples) && n_random_samples < nrow(param_grid)) {
+    param_grid = param_grid[sample(nrow(param_grid), n_random_samples), ]
+  }
+  return(param_grid)
+}
+
+
+compute_metric_per_trial = function(results_df, eval_function, params_df = NULL) {
+  #' Compute a scalar metric from each trial of an experiment.
+  #'
+  #' @param results_df A table of results, one row per trial
+  #' @param eval_function A function to evaluate each row, using the result and (optionally-null) parameters
+  #' @param params_df An optionally-null dataframe of paramters, one row per trial
+  #'
+  #' @return The list of metric values, one per row
+  #' The eval_function is run on the corresponding pairs of rows from results_df and params_df
+
+  # List of metrics from each run
+  metric_values = list()
+  # Iterate over all trials and compute the metric for each
+  for (i in seq_len(nrow(results_df))) {
+    # Get the shift predicted by the registration method
+    results = results_df[i, ]
+    # Get the map parameters defining the initial shift
+    if (is.null(params_df)) {
+      params = NULL
+    } else {
+      params = params_df[i, ]
+    }
+    # Compute a metric evaluating how well the predicted value matched the real one
+    metric = eval_function(results, params)
+    # Record this result
+    metric_values[[i]] = metric
+  }
+  return(metric_values)
+}
+
+visualize_metric_vs_1_parameter = function(
+    results_df,
+    parameter_df,
+    attributes_to_visualize,
+    eval_function,
+    smoothing_bandwidth = 5,
+    title = "") {
+  #' Visualize one metric versus one or more paramter values
+  #'
+  #' @param results_df The results of trials, one per row
+  #' @param paramter_df The paramters used to generate the results, one row per trial
+  #' @param attributes_to_visualize Which paramters to visualize the metrics versus
+  #' @param eval_function How to evalute the quality of the result
+  #' @param smoothing_bandwidth Width of the kernel used to smooth the trend line. May be a scalar or one value per attribute being visualized.
+  #' @param title Title of the plot
+
+
+  # Compute the metric values for each row in the results
+  metric_values = compute_metric_per_trial(
+    results_df = results_df,
+    eval_function = eval_function,
+    params_df = parameter_df
+  )
```

Comment:
Because the evaluation is being done per trial, the `eval_res_threshold` function always returns either 0 or 1, so it does not yield any insight beyond what `eval_res_dist` returns. For that to be useful, I think this evaluation would have to process multiple "trials" (random map generations using the same parameters) at once so that a cross-trial "success rate" could be computed.

Reply:
Smart, I agree the multi-trial approach would be better. I'd argue though that you get something close to a success rate using the kernel-smoothed trend line, since it's averaging the `0/1` values. But the individual data points are not particularly informative.

Reply:
OK fair point -- fine to keep as is for now!